library/core/src/ffi/c_str.rs

```diff
@@ -755,6 +760,39 @@ impl AsRef<CStr> for CStr {
 #[unstable(feature = "cstr_internals", issue = "none")]
 #[cfg_attr(bootstrap, rustc_const_stable(feature = "const_cstr_from_ptr", since = "1.81.0"))]
 #[rustc_allow_const_fn_unstable(const_eval_select)]
+// Preconditions:
+// - ptr must be non-null
+// - ptr must point to a valid null-terminated string
+// - null terminator must be within isize::MAX bytes from ptr
+#[requires(!ptr.is_null() && {
+    let mut found_null = false;
+    let mut valid = true;
+    for i in 0..isize::MAX as usize {
+        if unsafe { *ptr.add(i) } == 0 {
+            if found_null {
+                // Already found a null before, so this is an extra null
+                valid = false;
+                break;
+            }
+            found_null = true;
+            break;
+        }
+    }
+    found_null && valid
+})]
+// Postconditions:
+// - Returns length before null terminator
+// - Length must be less than isize::MAX
+#[ensures(|&result| result < isize::MAX as usize && {
+    let mut valid = true;
+    for i in 0..result {
+        if unsafe { *ptr.add(i) } == 0 {
+            valid = false;
+            break;
+        }
+    }
+    valid && unsafe { *ptr.add(result) } == 0
+})]
```

Comment:
I think you can simplify this a bit since we mostly care about safety. You could use `[slice_from_raw_parts](https://doc.rust-lang.org/std/ptr/fn.slice_from_raw_parts.html)` to create a slice ptr, check that we can dereference it, and ensure the last byte is 0.

Reply:
Sure. Updated the code.